### ISSUE TITLE
Race condition and possible heap corruption

### ISSUE THREAD

**Issue by pierotofy** [simonfuhrmann/mve#476]
Symptoms: sporadically `dmrecon` will crash with the following: https://gist.github.com/pierotofy/c49447e86a187e8ede50fb302cf5a47b

I've been able to get a core dump and the stack trace fails at https://github.com/simonfuhrmann/mve/blob/master/libs/dmrecon/patch_sampler.cc#L109 which is really odd because there's nothing that would suggest a problem around that line, and perhaps the presence of some odd memory corruption. 

```
(gdb) frame 13
#13 mvs::PatchSampler::fastColAndDeriv (this=<optimized out>, 
    v=<optimized out>, color=std::vector of length 0, capacity 0, 
    deriv=std::vector of length 0, capacity 0) at patch_sampler.cc:109
109	    std::vector<math::Vec2f> gradDir(nrSamples);
(gdb) info locals
refV = std::shared_ptr (count 3, weak 0) 0x0
nfp = <optimized out>
ratio = <optimized out>
mmLevel = <optimized out>
p1 = {static dim = <optimized out>, v = {698.763428, 431.70462, 106.666252}}
img = std::shared_ptr (count 2, weak 0) 0x0
gradDir = std::vector of length 0, capacity 0
(gdb) 
```

Here's the full core dump stack.

```
#0  0x00007fa2b9061428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007fa2b906302a in __GI_abort () at abort.c:89
#2  0x00007fa2b90a37ea in __libc_message (do_abort=do_abort@entry=2, 
    fmt=fmt@entry=0x7fa2b91bced8 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007fa2b90ae970 in malloc_printerr (ar_ptr=0x7fa2b93f0b20 <main_arena>, 
    ptr=0x1706c30, str=0x7fa2b91b9c92 "corrupted double-linked list", 
    action=<optimized out>) at malloc.c:5006
#4  _int_malloc (av=av@entry=0x7fa2b93f0b20 <main_arena>, 
    bytes=bytes@entry=200) at malloc.c:3727
#5  0x00007fa2b90b0184 in __GI___libc_malloc (bytes=200) at malloc.c:2913
#6  0x00007fa2b9de1e78 in operator new(unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x000000000042553a in __gnu_cxx::new_allocator<double>::allocate (
    this=<optimized out>, __n=<optimized out>)
    at /usr/include/c++/5/ext/new_allocator.h:104
#8  std::allocator_traits<std::allocator<double> >::allocate (__a=..., 
    __n=<optimized out>) at /usr/include/c++/5/bits/alloc_traits.h:491
#9  std::_Vector_base<double, std::allocator<double> >::_M_allocate (
    this=<optimized out>, __n=<optimized out>)
    at /usr/include/c++/5/bits/stl_vector.h:170
#10 std::_Vector_base<math::Vector<float, 2>, std::allocator<math::Vector<float,---Type <return> to continue, or q <return> to quit---
 2> > >::_M_create_storage (__n=<optimized out>, this=0x7ffcfcf8dd40)
    at /usr/include/c++/5/bits/stl_vector.h:185
#11 std::_Vector_base<math::Vector<float, 2>, std::allocator<math::Vector<float, 2> > >::_Vector_base (__a=..., __n=<optimized out>, this=0x7ffcfcf8dd40)
    at /usr/include/c++/5/bits/stl_vector.h:136
#12 std::vector<math::Vector<float, 2>, std::allocator<math::Vector<float, 2> > >::vector (__a=..., __n=<optimized out>, this=0x7ffcfcf8dd40)
    at /usr/include/c++/5/bits/stl_vector.h:278
#13 mvs::PatchSampler::fastColAndDeriv (this=<optimized out>, 
    v=<optimized out>, color=std::vector of length 0, capacity 0, 
    deriv=std::vector of length 0, capacity 0) at patch_sampler.cc:109
#14 0x000000000041cb32 in mvs::PatchOptimization::optimizeDepthOnly (
    this=this@entry=0x7ffcfcf8e180) at patch_optimization.cc:277
#15 0x000000000041d9d0 in mvs::PatchOptimization::doAutoOptimization (
    this=this@entry=0x7ffcfcf8e180) at patch_optimization.cc:178
#16 0x00000000004116d1 in mvs::DMRecon::processFeatures (
    this=this@entry=0x7ffcfcf8f4a0) at dmrecon.cc:295
#17 0x00000000004142ad in mvs::DMRecon::start (this=this@entry=0x7ffcfcf8f4a0)
    at dmrecon.cc:98
#18 0x000000000040ae90 in reconstruct (
    scene=std::shared_ptr (count 507, weak 0) 0x16f8c80, settings=...)
    at dmrecon.cc:63
#19 0x000000000040c167 in main._omp_fn.0(void) () at dmrecon.cc:311
---Type <return> to continue, or q <return> to quit---
#20 0x00007fa2b9834cbf in GOMP_parallel ()
   from /usr/lib/x86_64-linux-gnu/libgomp.so.1
#21 0x00000000004083ca in main (argc=<optimized out>, argv=<optimized out>)
    at dmrecon.cc:285
```

In the process of trying to fix this, I've added some "usleep" statements around dmrecon's code to help trigger this problem more consistently: https://github.com/pierotofy/mve/tree/segfaultpoc

With the changes in the segfaultpoc branch I can trigger a segfault more frequently (not always, takes a few runs). However, I don't know if this is the same segfault.

`OMP_NUM_THREADS=20 /path/to/mve/apps/dmrecon/dmrecon -s5 --progress=silent --local-neighbors=2 /path/to/mve_project`

```
Received signal SIGSEGV (segmentation fault)
Obtained 9 stack frames: 0x559aeabc5ca4 0x7f00491e1e00 0x559aeab89cbe 0x559aeab8a5b0 0x559aeab65139 0x7f00493b8fa3 0x559aeab620b1 0x7f00491ce223 0x559aeab62ede
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x77ca4)[0x559aeabc5ca4]
/usr/lib/libc.so.6(+0x37e00)[0x7f00491e1e00]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3bcbe)[0x559aeab89cbe]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3c5b0)[0x559aeab8a5b0]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x17139)[0x559aeab65139]
/usr/lib/libgomp.so.1(GOMP_parallel+0x43)[0x7f00493b8fa3]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x140b1)[0x559aeab620b1]
/usr/lib/libc.so.6(__libc_start_main+0xf3)[0x7f00491ce223]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x14ede)[0x559aeab62ede]
Segmentation fault
```

Oddly enough, I don't get a core dump (checked ulimit and presence of debug symbols). I can get core dumps in other programs. This stack trace looks different than the one I was trying to fix, but is identical to one that was reported here: https://github.com/simonfuhrmann/mve/issues/323

This happens with a variety of datasets (large or small) so the input project should not matter.

I spent a few days looking into this, without a solution. I think that perhaps the code in `ImagePyramidCache::get` and `ImagePyramidCache::cleanup()`, even though they have a lock, might have a role in this. The lock will get released as soon as the function scope exits, leaving code such as:

```
SingleView::loadColorImage(int _minLevel)
{
    minLevel = _minLevel;
    img_pyramid = ImagePyramidCache::get(this->scene, this->view, this->embedding, minLevel);
}
```

Possibly vulnerable to a race condition? (say `get` is called, `get` function returns, `cleanup` is called from another thread, stuff is deallocated, `img_pyramid` from first thread is assigned the value from `get`, but stuff has been deallocated).

Sorry for the length of this issue, I just thought I'd write down my findings. If you have any pointers that could help guide me in the direction for a fix please let me know? :pray:

**Comment by pierotofy** [comment]
Valgrind gives some more information:

```
==1890== Thread 15:
==1890== Invalid read of size 4
==1890==    at 0x143C83: mve::View::save_image_intern(mve::View::ImageProxy*) (view.cc:923)
==1890==    by 0x1445AF: mve::View::save_view() (view.cc:254)
==1890==    by 0x11F138: main._omp_fn.0 (dmrecon.cc:312)
==1890==    by 0x4D32ABD: gomp_thread_start (team.c:120)
==1890==    by 0x4D6DA9C: start_thread (in /usr/lib/libpthread-2.28.so)
==1890==    by 0x4E82B22: clone (in /usr/lib/libc-2.28.so)
==1890==  Address 0xc is not stack'd, malloc'd or (recently) free'd
==1890== 
Received signal SIGSEGV (segmentation fault)
Obtained 8 stack frames: 0x17fca4 0x4dbee00 0x143c83 0x1445b0 0x11f139 0x4d32abe 0x4d6da9d 0x4e82b23
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x77ca4)[0x17fca4]
/usr/lib/libc.so.6(+0x37e00)[0x4dbee00]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3bc83)[0x143c83]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3c5b0)[0x1445b0]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x17139)[0x11f139]
/usr/lib/libgomp.so.1(+0x16abe)[0x4d32abe]
/usr/lib/libpthread.so.0(+0x7a9d)[0x4d6da9d]
/usr/lib/libc.so.6(clone+0x43)[0x4e82b23]
Segmentation fault
```

**Comment by pierotofy** [comment]
A few more findings. After adding some print statements I've narrowed down a problem between calls to `View::cache_cleanup` and `View::save_image_intern`. In essence, `save_image_intern` seems to be running fine, until at a certain point the pointer to `proxy->image` turns to `0` in the middle of the execution (randomly between one of these statements):

```
    proxy->is_dirty = false;
    proxy->filename = filename;
    proxy->width = proxy->image->width();
    usleep(10000 + rand() % 50000);
    std::cerr << "G " << proxy->image << std::endl;
    proxy->height = proxy->image->height();
    proxy->channels = proxy->image->channels();
    proxy->type = proxy->image->get_type();
    proxy->is_initialized = true;
```

```
ipc_cleanup 1:end
ipc_cleanup 2:start
cache_cleanup: start
cache_cleanup: end
ipc_cleanup 2:end
ipc_cleanup 3:start
cache_cleanup: start
cache_cleanup: end
ipc_cleanup 3:end
ipc_cleanup 5:start
cache_cleanup: start
G 0
Received signal SIGSEGV (segmentation fault)
Obtained 8 stack frames: 0x55928117d054 0x7f555e2cfe00 0x559281140fea 0x5592811416ca 0x55928111c199 0x7f555e4afabe 0x7f555e463a9d 0x7f555e393b23
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x78054)[0x55928117d054]
/usr/lib/libc.so.6(+0x37e00)[0x7f555e2cfe00]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3bfea)[0x559281140fea]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x3c6ca)[0x5592811416ca]
/home/piero/Documents/mve/apps/dmrecon/dmrecon(+0x17199)[0x55928111c199]
/usr/lib/libgomp.so.1(+0x16abe)[0x7f555e4afabe]
/usr/lib/libpthread.so.0(+0x7a9d)[0x7f555e463a9d]
/usr/lib/libc.so.6(clone+0x43)[0x7f555e393b23]
Segmentation fault
```

This takes a few tries to segfault (it runs normally sometimes).

**Comment by simonfuhrmann** [comment]
Closed by #478
